### PR TITLE
Noun db

### PR DESCRIPTION
**Pull request by Ezibenroc**
Improve class Nounificator:
* No more 0/1 codes... A function per map instead (I kept the old function for code factoring purpose, but they should not be used outside of the class).
* Docstrings.

### CODE REVIEW

**Comment by Ezibenroc** [comment]
@ProgVal: I try to have a good programming style. Could you please check this pull request (especially the class [Nounificator](https://github.com/ProjetPP/PPP-QuestionParsing-Grammatical/blob/88e8fe6640aab3c17456f1bae444c7afaad15d97/ppp_questionparsing_grammatical/nounDB.py#L3)) and tell me if you see some weird things that should be changed?

**Comment by yhamoudi** [comment]
i would be nice to have shortest names for the functions (otherwise it's painful to use them)

**Comment by yhamoudi** [comment]
Why not: `D` instead of `Direct` and `I` instead of `Inverse` ("reverse" needs to be replaced by "inverse", i didn't know that "reverse" was still used in this file)

**Comment by Ezibenroc** [comment]
> i would be nice to have shortest names for the functions (otherwise it's painful to use them)

See [this comment](https://github.com/ProjetPP/PPP-QuestionParsing-Grammatical/pull/4#discussion_r19040092) of @Tpt at the beginning of the project :)

**Comment by yhamoudi** [comment]
Can we have autocompletion when we update the database from the terminal?

**Comment by Ezibenroc** [comment]
> Can we have autocompletion when we update the database from the terminal?

```
sudo aptitude install ipython3
ipython3
```

Enjoy!

**Comment by yhamoudi** [comment]
Cool! So just change "reverse" to "inverse"

File: ppp_questionparsing_grammatical/nounDB.py [hunk 1]
```diff
@@ -23,21 +26,27 @@ def __eq__(self, other):
...
         f.close()
 
-    def add(self, verb, noun, n):
-        if n==0:
+    def _add(self, verb, noun, mapChoice):
+        if mapChoice == self.direct:
             target = self.verbToNounsDirect
-        else:
+        elif mapChoice == self.reverse:
             target = self.verbToNounsReverse
```
Comment (ProgVal) [inline ppp_questionparsing_grammatical/nounDB.py]:
```
else:
    raise AssertionError('mapChoice should be self.direct or self.reverse, not %r' % mapChoice')
```

File: ppp_questionparsing_grammatical/nounDB.py [hunk 2]
```diff
@@ -45,41 +54,105 @@ def add(self, verb, noun, n):
...
+        """
+        for t in nounList:
+            self.addReverse(verb, t)
+
+    def _remove(self, verb, noun, mapChoice):
+        if mapChoice == self.direct:
             target = self.verbToNounsDirect
-        else:
+        elif mapChoice == self.reverse:
             target = self.verbToNounsReverse
```
Comment (ProgVal) [inline ppp_questionparsing_grammatical/nounDB.py]: same comment as above

File: ppp_questionparsing_grammatical/nounDB.py [hunk 3]
```diff
@@ -45,41 +54,105 @@ def add(self, verb, noun, n):
...
+            Remove all the reverse nounifications of the given verb.
+        """
+        self._removeVerb(verb, self.reverse)
+
+    def _toNouns(self, verb, mapChoice):
+        if mapChoice == self.direct:
             target = self.verbToNounsDirect
-        else:
+        elif mapChoice == self.reverse:
             target = self.verbToNounsReverse
```
Comment (ProgVal) [inline ppp_questionparsing_grammatical/nounDB.py]: same comment

File: ppp_questionparsing_grammatical/nounDB.py [hunk 4]
```diff
@@ -45,41 +54,105 @@ def add(self, verb, noun, n):
...
         try:
             return target[verb]
         except KeyError:
             return []
```
Comment (ProgVal) [inline ppp_questionparsing_grammatical/nounDB.py]: Replace these four lines by: `return target.get(verb, [])`

File: ppp_questionparsing_grammatical/nounDB.py (hunk 1 above)
Comment (Ezibenroc) [inline ppp_questionparsing_grammatical/nounDB.py]: Even if this function do not have to be used outside of the class?

Comment (ProgVal) [inline ppp_questionparsing_grammatical/nounDB.py]: Actually, couldn't the target be given as the parameter?

**Comment by ProgVal** [comment]
```
def __str__(self):
    l = sorted([(x, 0) for x in self.verbToNounsDirect.keys()] + [(x, 1) for x in self.verbToNounsReverse.keys()])
    return '\n'.join([self.select(x) for x in sorted(l)])
```

l is sorted twice

File: ppp_questionparsing_grammatical/nounDB.py (hunk 1 above)
Comment (Ezibenroc) [inline ppp_questionparsing_grammatical/nounDB.py]: Very good remark!

**Comment by yhamoudi** [comment]
Do we merge?

**Pull request merged**